dm/syncer/data_validator.go

```diff
@@ -15,23 +15,73 @@ package syncer
 
 import (
 	"context"
+	"fmt"
+	"hash/fnv"
+	"strings"
 	"sync"
 	"time"
 
 	"github.com/go-mysql-org/go-mysql/replication"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb-tools/pkg/filter"
+	"github.com/pingcap/tidb/parser/model"
+	"go.uber.org/atomic"
 	"go.uber.org/zap"
 
 	"github.com/pingcap/tiflow/dm/dm/config"
 	"github.com/pingcap/tiflow/dm/dm/pb"
 	"github.com/pingcap/tiflow/dm/dm/unit"
+	"github.com/pingcap/tiflow/dm/pkg/binlog"
+	"github.com/pingcap/tiflow/dm/pkg/binlog/event"
 	"github.com/pingcap/tiflow/dm/pkg/conn"
 	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
 	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/dm/pkg/schema"
 	"github.com/pingcap/tiflow/dm/pkg/terror"
 	"github.com/pingcap/tiflow/dm/pkg/utils"
 	"github.com/pingcap/tiflow/dm/syncer/dbconn"
 )
 
+const (
+	checkInterval      = 5 * time.Second
+	validationInterval = 10 * time.Second
+)
+
+type validateTableInfo struct {
+	Source     *filter.Table
+	Info       *model.TableInfo
+	PrimaryKey *model.IndexInfo
+	Target     *filter.Table // target table after route
+	pkIndices  []int
+}
+
+type rowChangeType int
+
+const (
+	rowInsert rowChangeType = iota
+	rowDeleted
+	rowUpdated
+)
+
+// change of table
+// binlog changes are clustered into table changes
+// the validator validates changes of table-grain at a time.
+type tableChange struct {
+	table *validateTableInfo
+	rows  map[string]*rowChange
+}
+
+// change of a row.
+type rowChange struct {
```

Comment:
please take a look at https://github.com/pingcap/tiflow/blob/master/pkg/sqlmodel/row_change.go

this library can be shared by DM and CDC, so maybe we can easier port data validator to CDC in future

Reply:
struct is similar,  with no `prev data` and more extra fields. will stick to it now, there may other changes later.

> BTW, does sync-diff-inspector from tidb-tools have some well-tested functions that you can use?

the data compare part comes from `sync-diff-inspector`, but this logic can be mush simplier in continuous validation, so it's reimplemented, most other code is continuous validation only